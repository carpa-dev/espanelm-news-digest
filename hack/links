https://brasil.elpais.com/brasil/2019/09/12/internacional/1568322473_115146.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/13/politica/1568373825_307103.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/13/ciencia/1568367630_243262.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/13/internacional/1568330740_731392.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/13/politica/1568331156_959924.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/13/opinion/1568332061_508519.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/12/cultura/1568301039_673770.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/10/internacional/1568126943_481320.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/12/opinion/1568307477_178549.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/12/deportes/1568244415_227393.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/12/politica/1568294834_398592.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/12/opinion/1568300730_780955.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/11/opinion/1568219236_285570.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/10/opinion/1568074359_593557.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/09/opinion/1568042120_747554.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/11/opinion/1568239079_683810.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/12/politica/1568245547_408819.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/10/politica/1568147928_339038.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/10/politica/1568134128_017016.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/11/politica/1568158371_184921.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/11/politica/1568163025_157063.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/10/deportes/1568150138_210957.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/12/internacional/1568241951_582237.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/11/internacional/1568223622_700924.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/10/internacional/1568134696_328590.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/11/internacional/1568234486_343717.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/08/19/cultura/1566230138_634355.html#?ref=rss&format=simple&link=link
https://elpais.com/brasil/2019/09/06/cultura/1567786560_937893.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/06/cultura/1567775581_186801.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/02/cultura/1567451707_224689.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/08/cultura/1567897264_809021.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/07/politica/1567812673_810000.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/09/politica/1568057092_806142.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/06/internacional/1567786813_480794.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/08/30/opinion/1567166223_815812.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/11/internacional/1568209720_396533.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/10/estilo/1568128681_462749.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/10/videos/1568127666_060077.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/10/economia/1568149947_521077.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/09/internacional/1568038662_788857.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/11/ciencia/1568217140_930068.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/09/ciencia/1568041613_749340.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/12/estilo/1568244284_221813.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/10/tecnologia/1568139357_684716.html#?ref=rss&format=simple&link=link
https://brasil.elpais.com/brasil/2019/09/11/tecnologia/1568189214_927431.html#?ref=rss&format=simple&link=link
https://elpais.com/internacional/2019/09/12/colombia/1568322473_115146.html
https://elpais.com/elpais/2019/09/13/ciencia/1568367630_243262.html
https://elpais.com/internacional/2019/09/13/estados_unidos/1568330740_731392.html
https://elpais.com/cultura/2019/09/12/actualidad/1568301039_673770.html
https://elpais.com/elpais/2019/09/10/planeta_futuro/1568126943_481320.html
https://elpais.com/internacional/2019/09/10/actualidad/1568134128_017016.html
https://elpais.com/internacional/2019/09/12/estados_unidos/1568241951_582237.html
https://elpais.com/internacional/2019/09/11/estados_unidos/1568223622_700924.html
https://elpais.com/sociedad/2019/09/10/actualidad/1568134696_328590.html
https://elpais.com/cultura/2019/09/08/television/1567897264_809021.html
https://elpais.com/sociedad/2019/09/06/actualidad/1567786813_480794.html
https://elpais.com/elpais/2019/08/30/ideas/1567166223_815812.html
https://elpais.com/elpais/2019/09/11/mamas_papas/1568209720_396533.html
https://elpais.com/elpais/2019/09/10/videos/1568127666_060077.html
https://elpais.com/elpais/2019/09/11/ciencia/1568217140_930068.html
https://elpais.com/elpais/2019/09/09/ciencia/1568041613_749340.html
https://elpais.com/tecnologia/2019/09/10/actualidad/1568139357_684716.html
https://elpais.com/tecnologia/2019/09/11/actualidad/1568189214_927431.html
