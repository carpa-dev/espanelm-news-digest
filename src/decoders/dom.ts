import { JSDOM } from 'jsdom';
import { Result } from 'true-myth';

export interface DOM {
  querySelector(q: string): null | Element;
  querySelectorAll(q: string): NodeListOf<Element>;
}

export function DOMFromText(s: string): Result<DOM, { reason: string }> {
  try {
    return Result.ok(new JSDOMImpl(s));
  } catch (error) {
    return Result.err({ reason: error });
  }
}

// JSDOM does some runtime to guarantee
// it's running inside a class
// therefore we need this wrapper
export class JSDOMImpl implements DOM {
  _dom: JSDOM;

  constructor(html: string) {
    this._dom = new JSDOM(html);
  }

  querySelector(q: string): Element | null {
    return this._dom.window.document.querySelector(q);
  }

  querySelectorAll(q: string): NodeListOf<Element> {
    return this._dom.window.document.querySelectorAll(q);
  }
}
