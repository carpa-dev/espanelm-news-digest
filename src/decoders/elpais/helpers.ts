import { DOM } from '../dom';
import { Maybe } from 'true-myth';

export function array<K>(dom: DOM, q: string): Maybe<K[]> {
  const body = Array.from(dom.querySelectorAll(q))
    .map(b => b.textContent || '')
    .filter(b => b.length > 1);

  if (body.length) {
    return Maybe.of(body);
  }
  return Maybe.nothing();
}
