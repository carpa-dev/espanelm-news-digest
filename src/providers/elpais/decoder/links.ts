import { DOM } from '../../../decoders/dom';
import { Maybe } from 'true-myth';

export interface PageLinks {
  ptbr: string;
  es: string;
}

export function decodeLinks(dom: DOM): Maybe<PageLinks> {
  // TODO: make it more functional

  const links = Array.from(dom.querySelectorAll('link[rel="alternate"]'));

  const extracted = {
    ptbr: filterLinks(links, 'pt-br') || filterLinks(links, 'pt-BR'),
    es: filterLinks(links, 'es') || filterLinks(links, 'es-ES')
  };

  if (extracted.ptbr && extracted.es) {
    return Maybe.just(extracted);
  }
  return Maybe.nothing();
}

function filterLinks(elements: Element[], lang: string): string {
  const el = elements.filter(isHTMLLink).find(a => a.hreflang === lang);

  if (el) {
    return el.href;
  }

  return '';
}

function isHTMLLink(e: Element): e is HTMLLinkElement {
  const asHTMLLink = e as HTMLLinkElement;

  if (asHTMLLink.href && asHTMLLink.hreflang) {
    return true;
  } else {
    return false;
  }
}
