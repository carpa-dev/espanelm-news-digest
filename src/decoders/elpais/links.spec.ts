import { loadFixtureOrFail } from '../../../test/helpers';
import { DOM, DOMFromText } from '../dom';
import { PageLinks, decodeLinks } from './links';
import { Result } from 'true-myth';

fdescribe('ElPais Links Decoder', () => {
  const domFromFixture = async (f: string): Promise<DOM> => {
    const fixture = await loadFixtureOrFail(f);

    return Result.unsafelyUnwrap(DOMFromText(fixture));
  };

  const unwrapped = (dom: DOM): PageLinks => decodeLinks(dom).unsafelyUnwrap();

  it('should work', async () => {
    const dom = await domFromFixture('article.html');

    expect(unwrapped(dom)).toMatchSnapshot();
  });

  it('should work with method 2', async () => {
    const dom = await domFromFixture('articles_es/1575633540_632181');

    expect(unwrapped(dom)).toMatchSnapshot();
  });

  it('should return Nothing if its unable to decode', async () => {
    const dom = DOMFromText('').unsafelyUnwrap();
    const body = decodeLinks(dom);
    expect(body.isNothing()).toBe(true);
  });
});
