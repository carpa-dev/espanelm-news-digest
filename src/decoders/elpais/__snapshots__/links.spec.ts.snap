// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElPais Links Decoder should work 1`] = `
Object {
  "es": "https://elpais.com/elpais/2019/09/10/videos/1568127666_060077.html",
  "ptbr": "https://brasil.elpais.com/brasil/2019/09/10/videos/1568127666_060077.html",
}
`;

exports[`ElPais Links Decoder should work with method 2 1`] = `
Object {
  "es": "https://elpais.com/internacional/2019/12/06/actualidad/1575633540_632181.html",
  "ptbr": "https://brasil.elpais.com/brasil/2019/12/06/internacional/1575633540_632181.html",
}
`;
